File: Android/app/src/main/java/com/caffeine/swingit/BonusGenerator.kt
package com.caffeine.swingit

import com.caffeine.swingit.Graphics.GInterval
import com.caffeine.swingit.Graphics.GPoint
import com.caffeine.swingit.Graphics.IGUpdatable

class BonusGenerator(val scene: GameScene) : IGUpdatable
{
    val screenHeightUsable: Float
    val positionInterval: GInterval
    var lastGenerationTime: Long = 0L

    init {
        screenHeightUsable = scene.size.height - scene.terrain.terrainHeightPixel
        positionInterval = GInterval(0f, screenHeightUsable)
    }


    override fun update(currentTime: Long)
    {
        if(currentTime - lastGenerationTime > scene.timelapseItemGeneration) {
            lastGenerationTime = currentTime
            scene.addChild(Bonus(scene,
                    GPoint(scene.size.width + scene.CHARACTER_SIZE.width,
                    positionInterval.random())
            ))
        }
    }
}

File: iOS/Swing-It/Swing-It/BonusGenerator.swift
//
//  BonusGenerator.swift
//  Swing-It
//
//  Created by Olivier Picard on 01/10/2018.
//  Copyright © 2018 Caffeine. All rights reserved.
//

import Foundation
import SpriteKit


class BonusGenerator : IUpdatable
{
    var scene: GameScene
    let minBorne: Int
    let maxBorne: Int
    var lastGenerationTime: TimeInterval = 0
    
    
    init(scene: GameScene)
    {
        self.scene = scene
        maxBorne = Int(scene.size.height - GameScene.bonusSize.height)
        minBorne = Int(scene.terrain.TERRAIN_POSITION + GameScene.bonusSize.height * 2)
    }
    
    
    func update(_ currentTime: TimeInterval) {
        if (currentTime - lastGenerationTime > scene.timelapseItemGeneration) {
            if (CGFloat.random(in: 0...1) > scene.BONUS_PROBABILITY) { return }
            lastGenerationTime = currentTime
            scene.addChild(Bonus(scene: scene,
                                 size: GameScene.bonusSize,
                                 position: CGPoint(x: Int(scene.size.width + GameScene.bonusSize.width),
                                                   y: Int.random(in: minBorne...maxBorne))))
        }
    }
}
